src/protocols/LoRaWAN/LoRaWAN.cpp

```diff
@@ -883,333 +986,378 @@ int16_t LoRaWANNode::activateABP(uint8_t initialDr) {
+  }
 
-  // calculate Time on Air of this uplink in milliseconds
-  this->lastToA = this->phyLayer->getTimeOnAir(uplinkMsgLen - RADIOLIB_LORAWAN_FHDR_LEN_START_OFFS) / 1000;
+  RadioLibTime_t tNow = mod->hal->millis();
+  // if scheduled uplink time is in the past, reschedule to now
+  if(this->tUplink < tNow) {
+    this->tUplink = tNow;
+  }
 
-  #if !RADIOLIB_STATIC_ONLY
-  delete[] uplinkMsg;
-  #endif
+  // if adhering to dutyCycle and the time since last uplink + interval has not elapsed, return an error
+  if(this->dutyCycleEnabled) {
+    if(this->rxDelayStart + (RadioLibTime_t)dutyCycleInterval(this->dutyCycle, this->lastToA) > this->tUplink) {
+      return(RADIOLIB_ERR_UPLINK_UNAVAILABLE);
+    }
+  }
+
+  // set the physical layer configuration for uplink
+  state = this->setPhyProperties(chnl,
+                                 RADIOLIB_LORAWAN_UPLINK, 
+                                 this->txPowerMax - 2*this->txPowerSteps);
   RADIOLIB_ASSERT(state);
   
-  // the downlink confirmation was acknowledged, so clear the counter value
-  this->confFCntDown = RADIOLIB_LORAWAN_FCNT_NONE;
-
-  // pass the extra info if requested
-  if(event) {
-    event->dir = RADIOLIB_LORAWAN_CHANNEL_DIR_UPLINK;
-    event->confirmed = isConfirmed;
-    event->confirming = isConfirmingDown;
-    event->datarate = this->dataRates[RADIOLIB_LORAWAN_CHANNEL_DIR_UPLINK];
-    event->freq = currentChannels[event->dir].freq;
-    event->power = this->txPowerMax - this->txPowerSteps * 2;
-    event->fCnt = this->fCntUp;
-    event->fPort = fPort;
+  // if requested, wait until transmitting uplink
+  tNow = mod->hal->millis();
+  if(this->tUplink > tNow) {
+    RADIOLIB_DEBUG_PROTOCOL_PRINTLN("Delaying transmission by %lu ms", (unsigned long)(this->tUplink - tNow));
+    if(this->tUplink > mod->hal->millis()) {
+      mod->hal->delay(this->tUplink - mod->hal->millis());
+    }
   }
 
-  // increase frame counter by one for the next uplink
-  this->fCntUp += 1;
+  state = this->phyLayer->transmit(in, len);
 
-  return(RADIOLIB_ERR_NONE);
+  // set the timestamp so that we can measure when to start receiving
+  this->rxDelayStart = mod->hal->millis();
+  RADIOLIB_DEBUG_PROTOCOL_PRINTLN("Uplink sent <-- Rx Delay start");
+
+  // increase Time on Air of the uplink sequence
+  this->lastToA += this->phyLayer->getTimeOnAir(len - RADIOLIB_LORAWAN_FHDR_LEN_START_OFFS) / 1000;
+
+  return(state);
+}
+
+// flag to indicate whether there was some action during Rx mode (timeout or downlink)
+static volatile bool downlinkAction = false;
+
+// interrupt service routine to handle downlinks automatically
+#if defined(ESP8266) || defined(ESP32)
+  IRAM_ATTR
+#endif
+static void LoRaWANNodeOnDownlinkAction(void) {
+  downlinkAction = true;
 }
 
-int16_t LoRaWANNode::downlinkCommon() {
+int16_t LoRaWANNode::receiveCommon(uint8_t dir, const LoRaWANChannel_t* dlChannels, const RadioLibTime_t* dlDelays, uint8_t numWindows, RadioLibTime_t tReference) {
```

Comment:
Is there a specific reason why `receiveCommon` has arguments? Both calls to this method use the same values (which are themselves members of `LoRaWANNode`, or a macro. Or is this needed for the relay, seeing as it is `protected`?

Reply:
Yup, re-used logic for derived classes. With different arguments there :)